### PR TITLE
8308452: Extend internal Architecture enum with byte order and address size

### PR DESCRIPTION
**Pull request by RogerRiggs**
The internal enum jdk.internal.util.Architecture does not provide information about the big or little endianness or the address size (64 or 32 bits).  The endian-ness and address size are intrinsic to the architecture.

The values of the enum are extended to separately identify the big endian and little-endian uses of the ISA.
For example, `PPC64` and `PPC64LE` for the big and little-endian versions.  
The enum values directly reflect the build-time artifacts and resulting executables.

This information about an architecture will make the enum more useful especially to identify a target platform in a cross-platform use case. A method is added to map well known aliases for the platforms to the Architecture enum.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8308452](https://bugs.openjdk.org/browse/JDK-8308452): Extend internal Architecture enum with byte order and address size (**Enhancement** - P4)

### Reviewers
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [2e04a1ca](https://git.openjdk.org/jdk/pull/14063/files/2e04a1cafb3b712fc2be1b3720b6022f78d670be)
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**)
 * [Amit Kumar](https://openjdk.org/census#amitkumar) (@offamitkumar - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14063/head:pull/14063` \
`$ git checkout pull/14063`

Update a local copy of the PR: \
`$ git checkout pull/14063` \
`$ git pull https://git.openjdk.org/jdk.git pull/14063/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14063`

View PR using the GUI difftool: \
`$ git pr show -t 14063`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14063.diff">https://git.openjdk.org/jdk/pull/14063.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14063#issuecomment-1561338738)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/util/Architecture.java [hunk 1]
```diff
@@ -33,16 +35,65 @@
...
-    AARCH64,
-    ARM,
-    RISCV64,
-    S390,
-    PPC64,
+    OTHER(0, null),          // An unknown architecture not specifically named
+    X64(64, ByteOrder.LITTLE_ENDIAN),  // Represents AMD64 and X86_64
+    X86(32, ByteOrder.LITTLE_ENDIAN),
+    AARCH64(64, ByteOrder.LITTLE_ENDIAN),
+    ARM(64, ByteOrder.LITTLE_ENDIAN),
```
Comment (TheRealMDoerr) [inline src/java.base/share/classes/jdk/internal/util/Architecture.java:42]: ARM should be 32 bit.

**Review by TheRealMDoerr — COMMENTED** [review]
I think this makes sense. Thanks for working on it! Most architectures have a fixed endianness.
For PPC64, I think using PPC64LE is fine to distinguish as it is a common name.
Another architecture which supports both endianness flavors is IA64. HPUX is big endian and linux little endian. I don't know how this should be handled. Fortunately, I don't think we need to support IA64. :-)

File: src/java.base/share/classes/jdk/internal/util/Architecture.java [hunk 2]
```diff
@@ -33,16 +35,65 @@
...
 public enum Architecture {
-    OTHER,      // An unknown architecture not specifically named
-    X64,        // Represents AMD64 and X86_64
-    X86,
-    AARCH64,
-    ARM,
-    RISCV64,
-    S390,
-    PPC64,
+    OTHER(0, null),          // An unknown architecture not specifically named
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/util/Architecture.java:38]: ByteOrder.nativeOrder() might be a better default for unknown archs.

File: src/java.base/share/classes/jdk/internal/util/OperatingSystem.java [hunk 3]
```diff
@@ -127,16 +127,25 @@ public static OperatingSystem current() {
...
+    }
+
+    /**
+     * {@return Simple conversion of ASCII to upper case; used when too early to use Locale conversions}
+     *
+     * @param str a string
+     * @return the string with lower case ascii converted to upper case
+     */
+    /* package-private */
+    static String toUpperCase(String str) {
```
Comment (AlanBateman) [inline src/java.base/share/classes/jdk/internal/util/OperatingSystem.java:140]: A bit icky to have OperatingSystem to expose a static method to convert Strings as it's nothing to do with OperatingSystem, is there anywhere else?

**Review by AlanBateman — COMMENTED** [review]
Mapping the arch to endianness seems okay for the current ports but I could imagine needing to re-visit that at some point.

**Comment by TheRealMDoerr** [comment]
I think the mapping from ppc64le to ppc64 should get removed with the introduction of ppc64le as separate architecture:
https://github.com/openjdk/jdk/blob/64f6681cc7bb184112adcf2b4f2c313e1b0a5c4f/src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Platform.java#LL55C5-L55C5
Would you mind removing it with your change?

File: src/java.base/share/classes/jdk/internal/util/OperatingSystem.java (hunk 3 above)
Comment (RogerRiggs) [reply src/java.base/share/classes/jdk/internal/util/OperatingSystem.java]: Delaying the initialization of OperatingSystem made it possible to revert to String.toUpperCase(). 
Enabled by the fix in PR [14181](https://github.com/openjdk/jdk/pull/14181).

**Review by TheRealMDoerr — COMMENTED** [review]
Thanks for the updates! Unfortunately, there's still something going wrong on ppc64le:
- jdk/internal/util/ArchTest.java: System property os.arch: "ppc64le", Architecture.current(): "PPC64"
- tools/jlink/plugins/CDSPluginTest.java: Cannot generate CDS archives: target image platform linux-ppc64le is different from runtime platform linux-ppc64

Do you happen to know why Architecture.current() misses the "le"?

**Comment by RogerRiggs** [comment]
> Do you happen to know why Architecture.current() misses the "le"?

@TheRealMDoerr  The makefile that does the substitution (in GensrcMisc.gmk) for the OPENJDK_TARGET_CPU is using the _CANONICAL form of the architecture (that maps ppc64le to ppc64).
I'll work up a fix; note that we do want the canonical form for s390x and amd/x86_64.

File: src/java.base/share/classes/jdk/internal/util/OperatingSystem.java [hunk 4]
```diff
@@ -26,6 +26,8 @@
 import jdk.internal.util.PlatformProps;
 import jdk.internal.vm.annotation.ForceInline;
 
+import java.util.Locale;
```
Comment (naotoj) [inline src/java.base/share/classes/jdk/internal/util/OperatingSystem.java:29]: Nit: Stray import?

**Review by TheRealMDoerr — APPROVED** [review]
Thanks for the updates! LGTM.

File: src/java.base/share/classes/jdk/internal/util/Architecture.java [hunk 5]
```diff
@@ -33,19 +35,81 @@
...
+    PPC64LE(64, ByteOrder.LITTLE_ENDIAN),
+    MIPSEL(32, ByteOrder.LITTLE_ENDIAN),
+    MIPS64EL(64, ByteOrder.LITTLE_ENDIAN)
     ;
 
+    private final int addrSize;
+    private final ByteOrder byteOrder;
+
+    /**
+     * Construct an Arctecture with number of address bits and byte order.
```
Comment (jaikiran) [inline src/java.base/share/classes/jdk/internal/util/Architecture.java:60]: Hello Roger, there is a typo here "Arctecture" instead of "Architecture"

**Review by jaikiran — APPROVED** [review]
The changes in this PR look OK to me. The introduction of instance method `byteOrder()` in the `Architecture` enum, as proposed in this PR, will help with the changes being proposed in https://github.com/openjdk/jdk/pull/11943. That way jlink won't have to have its own mapping of osname-architecture to endianness.
Once this PR is integrated, I'll update that other PR to use this new API and remove the endianness mappings being proposed in that PR.

**Comment by RogerRiggs** [comment]
I think this has settled down.
If anyone would care to test it on S390, PPC, LOONGARCH64, or MIPS that would be appreciated.

**Review by mlchung — APPROVED** [review]
This change looks fine to me.

**Comment by TheRealMDoerr** [comment]
Still good on PPC64.

**Review by TheRealMDoerr — APPROVED** [review]

**Review by offamitkumar — APPROVED** [review]
Looks good for s390.

**Pull request closed without merging**